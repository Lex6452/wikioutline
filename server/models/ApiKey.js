import { DataTypes, sequelize } from '../sequelize';
import randomstring from 'randomstring';

const ApiKey = sequelize.define(
  'apiKeys',
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    name: DataTypes.STRING,
    secret: { type: DataTypes.STRING, unique: true },
    userId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: 'users',
      },
    },
  },
  {
    tableName: 'apiKeys',
    paranoid: true,
    hooks: {
      beforeValidate: key => {
        key.secret = randomstring.generate(38);
      },
    },
  }
);

export default ApiKey;
