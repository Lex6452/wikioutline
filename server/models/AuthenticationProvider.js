// @flow
import providers from "../auth/providers";
import { DataTypes, sequelize } from "../sequelize";

const AuthenticationProvider = sequelize.define(
  "authentication_providers",
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    name: {
      type: DataTypes.STRING,
      validate: {
        isIn: [providers.map((p) => p.id)],
      },
    },
    enabled: {
      type: DataTypes.BOOLEAN,
      defaultValue: true,
    },
    providerId: {
      type: DataTypes.STRING,
    },
  },
  {
    timestamps: true,
    updatedAt: false,
  }
);

AuthenticationProvider.associate = (models) => {
  AuthenticationProvider.belongsTo(models.Team);
  AuthenticationProvider.hasMany(models.UserAuthentication);
};

export default AuthenticationProvider;
