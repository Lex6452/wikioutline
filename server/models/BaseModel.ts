import { Model, Column, PrimaryKey, IsUUID } from "sequelize-typescript";

class BaseModel extends Model {
  @IsUUID(4)
  @Column
  @PrimaryKey
  id: string;
}

export default BaseModel;
