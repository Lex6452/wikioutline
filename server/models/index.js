import User from './User';
import Team from './Team';
import Collection from './Collection';
import Document from './Document';
import Revision from './Revision';
import ApiKey from './ApiKey';

export { User, Team, Collection, Document, Revision, ApiKey };
