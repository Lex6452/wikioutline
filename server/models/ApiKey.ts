import randomstring from "randomstring";
import {
  Column,
  Table,
  Unique,
  DeletedAt,
  CreatedAt,
  UpdatedAt,
  BeforeValidate,
  BelongsTo,
} from "sequelize-typescript";
import BaseModel from "./BaseModel";
import User from "./User";

@Table({ tableName: "apiKeys", modelName: "apiKey" })
class ApiKey extends BaseModel {
  @Column
  name: string;

  @Column
  @Unique
  secret: string;

  @CreatedAt
  createdAt: Date;

  @UpdatedAt
  updatedAt: Date;

  @DeletedAt
  deletedAt: Date;

  // hooks

  @BeforeValidate
  static async generateSecret(instance: ApiKey) {
    if (!instance.secret) {
      instance.secret = randomstring.generate(38);
    }
  }

  // associations

  @BelongsTo(() => User, "userId")
  user: User;
}

export default ApiKey;
