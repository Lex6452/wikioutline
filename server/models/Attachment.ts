import path from "path";
import { FindOptions } from "sequelize";
import {
  BeforeDestroy,
  BelongsTo,
  Column,
  Default,
  ForeignKey,
  IsIn,
  Table,
} from "sequelize-typescript";
import { deleteFromS3, getFileByKey } from "@server/utils/s3";
import { DataTypes } from "../sequelize";
import Document from "./Document";
import Team from "./Team";
import User from "./User";
import BaseModel from "./base/BaseModel";

@Table({ tableName: "attachments", modelName: "attachment" })
class Attachment extends BaseModel {
  @Column
  key: string;

  @Column
  url: string;

  @Column
  contentType: string;

  @Column(DataTypes.BIGINT)
  size: number;

  @Column
  @Default("public-read")
  @IsIn([["private", "public-read"]])
  acl: string;

  get name() {
    return path.parse(this.key).base;
  }

  get redirectUrl() {
    return `/api/attachments.redirect?id=${this.id}`;
  }

  get isPrivate() {
    return this.acl === "private";
  }

  get buffer() {
    return getFileByKey(this.key);
  }

  // hooks

  @BeforeDestroy
  static async deleteAttachmentFromS3(model: Attachment) {
    await deleteFromS3(model.key);
  }

  // associations

  @BelongsTo(() => Team, "teamId")
  team: Team;

  @ForeignKey(() => Team)
  @Column
  teamId: string;

  @BelongsTo(() => Document, "documentId")
  document: Document;

  @ForeignKey(() => Document)
  @Column
  documentId: string;

  @BelongsTo(() => User, "userId")
  user: User;

  @ForeignKey(() => User)
  @Column
  userId: string;

  static findAllInBatches = async (
    query: FindOptions<Attachment>,
    callback: (
      attachments: Array<Attachment>,
      query: FindOptions<Attachment>
    ) => Promise<void>
  ) => {
    if (!query.offset) query.offset = 0;
    if (!query.limit) query.limit = 10;
    let results;

    do {
      results = await Attachment.findAll(query);
      await callback(results, query);
      query.offset += query.limit;
    } while (results.length >= query.limit);
  };
}

export default Attachment;
