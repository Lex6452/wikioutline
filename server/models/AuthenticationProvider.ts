import { Op } from "sequelize";
import {
  BelongsTo,
  Column,
  CreatedAt,
  DataType,
  Default,
  ForeignKey,
  HasMany,
  IsIn,
  Table,
} from "sequelize-typescript";
import { ValidationError } from "../errors";
import providers from "../routes/auth/providers";
import Team from "./Team";
import UserAuthentication from "./UserAuthentication";
import BaseModel from "./base/BaseModel";

@Table({
  tableName: "authentication_providers",
  modelName: "authentication_provider",
  timestamps: true,
})
class AuthenticationProvider extends BaseModel {
  @IsIn([providers.map((p) => p.id)])
  @Column
  name: string;

  @Default(true)
  @Column
  enabled: boolean;

  @Column
  providerId: string;

  @CreatedAt
  createdAt: Date;

  // associations

  @BelongsTo(() => Team, "teamId")
  team: Team;

  @ForeignKey(() => Team)
  @Column(DataType.UUID)
  teamId: string;

  @HasMany(() => UserAuthentication, "providerId")
  userAuthentications: UserAuthentication[];

  disable = async function () {
    const res = await AuthenticationProvider.findAndCountAll({
      where: {
        teamId: this.teamId,
        enabled: true,
        id: {
          [Op.ne]: this.id,
        },
      },
      limit: 1,
    });

    if (res.count >= 1) {
      return this.update({
        enabled: false,
      });
    } else {
      throw ValidationError("At least one authentication provider is required");
    }
  };

  enable = function () {
    return this.update({
      enabled: true,
    });
  };
}

export default AuthenticationProvider;
