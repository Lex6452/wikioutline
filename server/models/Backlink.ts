import { BelongsTo, ForeignKey, Column, Table } from "sequelize-typescript";
import Document from "./Document";
import User from "./User";
import BaseModel from "./base/BaseModel";

@Table({ tableName: "backlinks", modelName: "backlink" })
class Backlink extends BaseModel {
  @BelongsTo(() => User, "userId")
  user: User;

  @ForeignKey(() => User)
  @Column
  userId: string;

  @BelongsTo(() => Document, "documentId")
  document: Document;

  @ForeignKey(() => Document)
  @Column
  documentId: string;

  @BelongsTo(() => Document, "reverseDocumentId")
  reverseDocument: Document;

  @ForeignKey(() => Document)
  @Column
  reverseDocumentId: string;
}

export default Backlink;
