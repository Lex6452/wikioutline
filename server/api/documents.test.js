import TestServer from 'fetch-test-server';
import app from '..';
import { View } from '../models';
import { flushdb, seed } from '../test/support';

const server = new TestServer(app.callback());

beforeEach(flushdb);
afterAll(() => server.close());

describe('#documents.list', async () => {
  it('should return documents', async () => {
    const { user, document } = await seed();
    const res = await server.post('/api/documents.list', {
      body: { token: user.getJwtToken() },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.length).toEqual(2);
    expect(body.data[0].id).toEqual(document.id);
  });

  it('should allow changing sort direction', async () => {
    const { user, document } = await seed();
    const res = await server.post('/api/documents.list', {
      body: { token: user.getJwtToken(), direction: 'ASC' },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data[1].id).toEqual(document.id);
  });

  it('should require authentication', async () => {
    await seed();
    const res = await server.post('/api/documents.list');
    const body = await res.json();

    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });
});

describe('#documents.viewed', async () => {
  it('should return empty result if no views', async () => {
    const { user } = await seed();
    const res = await server.post('/api/documents.viewed', {
      body: { token: user.getJwtToken() },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.length).toEqual(0);
  });

  it('should return recently viewed documents', async () => {
    const { user, document } = await seed();
    await View.increment({ documentId: document.id, userId: user.id });

    const res = await server.post('/api/documents.viewed', {
      body: { token: user.getJwtToken() },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.length).toEqual(1);
    expect(body.data[0].id).toEqual(document.id);
  });

  it('should require authentication', async () => {
    await seed();
    const res = await server.post('/api/documents.viewed');
    const body = await res.json();

    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });
});
