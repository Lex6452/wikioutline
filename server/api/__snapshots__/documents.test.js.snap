// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#documents.list should require authentication 1`] = `
Object {
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#documents.viewed should require authentication 1`] = `
Object {
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;
